docs/dependencies.rst

```diff
@@ -0,0 +1,189 @@
+
+
+.. figure:: _static/img/test-deps.svg
+  :align: center
+  :alt: Simple test dependency presented conceptually.
+
+For most of the cases, this is sufficient to reason about test dependencies.
+In reality, as mentioned above, dependencies are handled at the level of test cases.
+Test cases on different partitions are always independent.
+If not specified differently, test cases with using programming environments are also independent.
+This is the default behavior of the :func:`depends_on` function.
+The following image shows the actual test case dependencies assuming that both tests support the ``E0`` and ``E1`` programming environments (for simplicity, we have omitted the partitions, since tests are always independent in that dimension):
+
+.. figure:: _static/img/test-deps-by-env.svg
+  :align: center
+  :alt: Test case dependencies by environment (default).
+
+This means that test cases of :class:`T1` may start executing before all test cases of :class:`T0` have finished.
+You can impose a stricter dependency between tests, such that :class:`T1` does not start execution unless all test cases of :class:`T0` have finished.
+You can achieve this as follows:
+
+.. code:: python
+
+   @rfm.simple_test
+   class T1(rfm.RegressionTest):
+       def __init__(self):
+           ...
+           self.depends_on('T0', how=rfm.DEPEND_FULLY)
+
+This will create the following test case graph:
+
+.. figure:: _static/img/test-deps-fully.svg
+  :align: center
+  :alt: Fully dependent test cases.
+
+
+You may also create arbitrary dependencies between the test cases of different tests, like in the following example, where the dependencies cannot be represented in any of the other two ways:
+
+.. figure:: _static/img/test-deps-exact.svg
+  :align: center
+  :alt: Arbitrary test cases dependencies
+
+These dependencies be achieved as follows:
+
+.. code:: python
+
+   @rfm.simple_test
+   class T1(rfm.RegressionTest):
+       def __init__(self):
+           ...
+           self.depends_on('T0', how=rfm.DEPEND_EXACT,
+                           subdeps={'E0': ['E0', 'E1'], 'E1': ['E1']})
+
+The ``subdeps`` argument defines the sub-dependencies between the test cases of :class:`T1` and :class:`T0` using an adjacency list representation.
+
+
+Cyclic dependencies
+-------------------
+
+Obviously, cyclic dependencies between test cases are not allowed.
+Cyclic dependencies between tests are not allowed either, even if the test case dependency graph is acyclic.
+For example, the following dependency set up is invalid:
+
+.. figure:: _static/img/test-deps-cycle.svg
+  :align: center
+  :alt: Any cyclic dependencies between tests are not allowed, even if the underlying test case dependencies are not forming a cycle.
+
+The test case dependencies here are clearly not forming a cycle, but the edge from ``(T0, E0)`` to ``(T1, E1)`` introduces a dependency from ``T0`` to ``T1`` forming a cycle at the test level.
```

Comment:
```suggestion
The test case dependencies here are clearly don't form a cycle, but the edge from ``(T0, E0)`` to ``(T1, E1)`` introduces a dependency from ``T0`` to ``T1`` forming a cycle at the test level.
```

Reply:
If you change this as Theo suggests, remove the `are` from `here are clearly don't form a cycle`.

Reply:
You probably mean to remove "are" here, too.